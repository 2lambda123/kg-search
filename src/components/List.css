/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

.kgs-list {
    position: relative;
    margin: 0 -20px;
}

@media (min-width: 576px) {
    .kgs-list {
        margin: 0;
    }
}

.kgs-list > ul {
    list-style-type: none;
    display: block;
    -ms-flex-flow: unset;
    flex-flow: unset;
    margin: 0;
    padding: 0;
}

.kgs-list.is-list > ul {
    display: block;
    -ms-flex-flow: unset;
    flex-flow: unset;
}

.kgs-list > div {
    border-top: 1px solid #777;
    width: 100%;
    margin-top: 10px;
    padding: 10px 0;
    font-size: var(--font-size-small);
    color: #777;
}

.kgs-list.is-list > div {
    margin-top: 40px;
    padding-bottom: 20px;
}

.kgs-list > ul > li {
    position: relative;
    display: inline-block;
    width: 100%;
    margin: 0;
    border-bottom: 1px solid black;
    background-color: white;
}

.kgs-list > ul > li>button {
    display: inline-block;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 20px 40px 20px 20px;
    border: 0;
    background-color: transparent;
    overflow: hidden;
}

.kgs-list > ul > li>button .kgs-list__chevron {
    position: absolute;
    top: calc(50% - 9px);
    right: 15px;
    color: rgb(108, 115, 121);
}

.kgs-list > ul > li>button:hover .kgs-list__chevron {
    color: rgb(51, 51, 51);
}

@media (min-width: 576px) {
    .kgs-list > ul > li {
        margin: 10px 0 10px 0;
        border-bottom-width: 0;
        -webkit-box-shadow: 2px 2px 4px#cbc9c9;
        box-shadow: 2px 2px 4px#cbc9c9;
        transition: box-shadow 0.2s ease-in;
    }

    .kgs-list > ul > li:hover {
        -webkit-box-shadow: 3px 3px 6px #8f8a8a;
        /* #cbc9c9; */
        box-shadow: 3px 3px 6px #8f8a8a;
        /* #cbc9c9; */
    }
}

@media (min-width: 1450px) {
    .kgs-list > ul {
        display: grid;
        grid-template-columns: repeat(auto-fill, minmax(440px, 1fr));
        grid-row-gap: 25px;
        -moz-row-gap: 25px;
        -webkit-row-gap: 25px;
        row-gap: 25px;
        grid-column-gap: 25px;
        -moz-column-gap: 25px;
        -webkit-column-gap: 25px;
        column-gap: 25px;
        margin: 0 0 25px 0;
        padding: 0;
    }

    .kgs-list.is-grid > ul > li {
        margin: 0;
        transition: box-shadow 0.2s ease-in, transform 0.2s ease-in;
    }

    .kgs-list.is-grid > ul > li:hover {
        transform: scale(1.05);
    }

    .kgs-list.is-list > ul > li {
        width: 100%;
        height: auto;
        height: unset;
        margin: 0 0 20px 0;
    }

    .kgs-list > ul > li {
        height: 440px;
        margin: 10px;
    }
}

/* Internet Explorer 10 & 11 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) and (min-width: 992px) {
    .kgs-list.is-grid > ul > li {
        margin: 10px 20px 10px 0;
    }
}
