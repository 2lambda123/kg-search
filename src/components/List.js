/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import PropTypes from "prop-types";
import "./List.css";

const ListItem = ({ reference, data, itemComponent, onClick }) => {

  const handleClick = (event) => {
    onClick(data, event.currentTarget);
  };

  const Component = itemComponent;
  return (
    <li>
      <button role="link" onClick={handleClick} data-reference={reference}>
        <Component data={data} />
        <span className="kgs-list__chevron"><i className="fa fa-chevron-right"></i></span>
      </button>
    </li>
  );
};

export const List = ({ className, title, items, itemComponent, layout, getKey, onClick }) => {
  if (!Array.isArray(items) || items.length === 0 || typeof getKey !== "function") {
    return null;
  }
  const classNames = ["kgs-list", className, layout === "grid" ? "is-grid" : "is-list"].join(" ");
  return (
    <div className={classNames}>
      {title && (
        <div>{title}</div>
      )}
      <ul>
        {items.map(item => {
          const key = getKey(item);
          return (
            <ListItem key={key} reference={key} data={item} itemComponent={itemComponent} onClick={onClick} />
          );
        })}
      </ul>
    </div>
  );
};

List.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string,
  items: PropTypes.arrayOf(PropTypes.any),
  itemComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object
  ]).isRequired,
  layout: PropTypes.string,
  getKey: PropTypes.func.isRequired,
  onClick: PropTypes.func
};

export default List;