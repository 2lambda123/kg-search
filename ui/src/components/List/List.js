/*
 *   Copyright (c) 2018, EPFL/Human Brain Project PCO
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import React from "react";

import "./List.css";
class Item extends React.PureComponent {

  handleClick = e => {
    const { item, onClick } = this.props;
    if (typeof onClick === "function") {
      e.stopPropagation();
      onClick(item);
    }
  }

  handleKeyDown = e => {
    const { item, onClick } = this.props;
    if (e.keyCode === 13 && typeof onClick === "function") {
      e.stopPropagation();
      onClick(item);
    }
  }

  setFocus = () => {
    const { hasFocus } = this.props;
    if (hasFocus) {
      this.ref.focus();
    }
  }

  componentDidMount() {
    this.setFocus();
  }

  componentDidUpdate() {
    this.setFocus();
  }

  render() {
    const { ItemComponent, item } = this.props;
    return (
      <div className="kgs-list-item" tabIndex="0" ref={ref => this.ref = ref} onClick={this.handleClick} onKeyDown={this.handleKeyDown} >
        <ItemComponent item={item} />
      </div>
    );
  }
}

export const List = ({ items, currentItem, ItemComponent, itemUniqKeyAttribute, onItemClick }) => {
  if (!Array.isArray(items) || !items.length) {
    return null;
  }
  return (
    <div className="kgs-list" >
      {items.map(item => <Item key={item[itemUniqKeyAttribute]} ItemComponent={ItemComponent} item={item} hasFocus={item === currentItem} onClick={onItemClick} />)}
    </div>
  );
};

export default List;