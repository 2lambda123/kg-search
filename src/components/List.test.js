/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import uniqueId from "lodash/uniqueId";
import renderer from "react-test-renderer";
import Enzyme, { mount, shallow, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import List from "./List";

Enzyme.configure({ adapter: new Adapter() });

test('List component renders initially', () => {
    const component = renderer.create(
        <List className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => null} />
    );
  
    expect(component.toJSON()).toMatchSnapshot();
});
  
test('List test className"', () => {
    const component = shallow(
        <List className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => null} />
    );
    expect(component.hasClass("className"));
});

test('List test title"', () => {
    const component = shallow(
        <List className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => null} />
    );
    expect(component.find("div div").text()).toBe("a title");
});

test('List test number of items', () => {
    const component = render(
        <List className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => null} />
    );
    expect(component.find("li").length).toBe(2);
});

test('Toggle test  button click', () => {
    const fn = jest.fn();
    const component = mount(
        <List className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={fn} itemComponent={() => null} />
    );
    component.find('button').at(1).simulate('click');
    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0].value).toBe("another value");
});