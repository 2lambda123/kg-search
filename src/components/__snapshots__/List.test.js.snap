// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders initially 1`] = `
<div
  className="kgs-list className is-list"
>
  <div>
    a title
  </div>
  <ul>
    <li>
      <button
        data-reference="1"
        onClick={[Function]}
        role="link"
      >
        <span
          className="kgs-list__chevron"
        >
          <i
            className="fa fa-chevron-right"
          />
        </span>
      </button>
    </li>
    <li>
      <button
        data-reference="2"
        onClick={[Function]}
        role="link"
      >
        <span
          className="kgs-list__chevron"
        >
          <i
            className="fa fa-chevron-right"
          />
        </span>
      </button>
    </li>
  </ul>
</div>
`;
